src/astarte-client/transforms/dataTree.ts

```diff
@@ -0,0 +1,346 @@
+          return new AstarteDataTreeLeaf<AstartePropertyData>({
+            interface: iface,
+            data: {
+              endpoint: `${endpoint}/${prop}`,
+              ...({
+                value: propValue,
+                type: getEndpointDataType(iface, `${endpoint}/${prop}`),
+              } as AstarteDataTuple),
+            },
+            endpoint: `${endpoint}/${prop}`,
+            relativeEndpoint: prop,
+            parentNode: this as AstarteDataTreeRoot<AstartePropertyData>,
+          });
+        }
+        if (
+          iface.type === 'datastream' &&
+          iface.aggregation === 'individual' &&
+          isIndividualDatastreamInterfaceValue(propValue)
+        ) {
+          const leafData: AstarteDatastreamIndividualData[] = [
+            {
+              endpoint: `${endpoint}/${prop}`,
+              timestamp: propValue.timestamp,
+              ...({
+                value: propValue.value,
+                type: getEndpointDataType(iface, `${endpoint}/${prop}`),
+              } as AstarteDataTuple),
+            },
+          ];
+          return new AstarteDataTreeLeaf<AstarteDatastreamIndividualData>({
+            interface: iface,
+            data: leafData,
+            endpoint: `${endpoint}/${prop}`,
+            relativeEndpoint: prop,
+            parentNode: this as AstarteDataTreeRoot<AstarteDatastreamIndividualData>,
+          });
+        }
+        if (
+          iface.type === 'datastream' &&
+          iface.aggregation === 'object' &&
+          isAggregatedDatastreamInterfaceValue(propValue)
+        ) {
+          const leafData: AstarteDatastreamObjectData[] = propValue.map((obj) => ({
+            endpoint: `${endpoint}/${prop}`,
+            timestamp: obj.timestamp,
+            value: Object.entries(_.omit(obj, 'timestamp')).reduce(
+              (acc, [objProp, objPropValue]) => ({
+                ...acc,
+                [objProp]: {
+                  value: objPropValue,
+                  type: getEndpointDataType(iface, `${endpoint}/${prop}/${objProp}`),
+                } as AstarteDataTuple,
+              }),
+              {},
+            ),
+          }));
+          return new AstarteDataTreeLeaf<AstarteDatastreamObjectData>({
+            interface: iface,
+            data: leafData,
+            endpoint: `${endpoint}/${prop}`,
+            relativeEndpoint: prop,
+            parentNode: this as AstarteDataTreeRoot<AstarteDatastreamObjectData>,
+          });
+        }
+        return new AstarteDataTreeRoot<Data>({
+          interface: iface,
+          data: propValue,
+          endpoint: `${endpoint}/${prop}`,
+          relativeEndpoint: prop,
+          parentNode: this,
+        });
+      })
+      .flat() as Array<AstarteDataTreeRoot<Data> | AstarteDataTreeLeaf<Data>>;
+  }
+
+  getNode(endpoint: string): AstarteDataTreeRoot<Data> | AstarteDataTreeLeaf<Data> | null {
+    const sanitizedEndpoint = endpoint.replace(/\/$/, '');
```

Comment:
you can save the name without the `/` and add it only when linearizing or when converting to string.
Saving it with a slash forces you to remove it everytime you have to compare/manipulate data

Reply:
Saving the full endpoint actually helps when comparing / manipulating data and the recursive search function becomes trivial.
Here I am doing `endpoint.replace(/\/$/, '')`  just to be able to find nodes both with `getNode('/sensors/gps')` and `getNode('/sensors/gps/')` (final slash).